vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -90,9 +93,13 @@ public void close() {
     MongoCollection<JsonObject> coll = getCollection(collection, writeOption);
     Object id = document.getValue(ID_FIELD);
     if (id == null) {
-      coll.insertOne(document, convertCallback(resultHandler, wr -> document.getString(ID_FIELD)));
+      coll.insertOne(document, convertCallback(resultHandler, wr -> useObjectId ? document.getJsonObject(ID_FIELD).getString(OID_FIELD) : document.getString(ID_FIELD)));
     } else {
-      coll.replaceOne(wrap(new JsonObject().put(ID_FIELD, document.getValue(ID_FIELD))), document, convertCallback(resultHandler, result -> null));
+      JsonObject filter = new JsonObject();
+      encodeKeyWhenUseObjectId(document);
```

Comment:
Ergh, yeah, there is no good option on this one, you *can* use an AtomicReference to get around the effectively final thing, but here its more trouble than its worth. 

It sucks that mongo is returning its ids by mutating its arguments in the first place, this has already bitten me once on other stuff I have been working on, I am tempted with all of the vertx-mongo-client stuff to do a copy of all of the arguments first so that we don't mutate objects that come in as args to avoid the issues I had.

You have 4 uses of encodeKeyWhenUseObjectId, only this one cant be fixed as the object is used in a lambda, so I would fix the others. 

You could assign to a new variable, i.e:

```
    JsonObject idEncodedDocument = encodeKeyWhenUseObjectId(document);

    MongoCollection<JsonObject> coll = getCollection(collection, writeOption);
    coll.insertOne(idEncodedDocument, convertCallback(resultHandler, wr -> {
      if (id) {
        return null;
      } else {
        JsonObject documentWithId = decodeKeyWhenUseObjectId(idEncodedDocument);
        return documentWithId.getString(ID_FIELD);
      }
    }));
    return this;
```

But it is up to you if you want to do that, I am not too bothered as mongo has already screwed us on this one.